homeassistant/components/assist_pipeline/pipeline.py

```diff
@@ -393,6 +421,141 @@ def end(self) -> None:
+            )
+        )
+
+        wake_word_settings = self.wake_word_settings or WakeWordSettings()
+
+        wake_word_vad: VoiceActivityTimeout | None = None
+        if (wake_word_settings.timeout is not None) and (
+            wake_word_settings.timeout > 0
+        ):
+            # Use VAD to determine timeout
+            wake_word_vad = VoiceActivityTimeout(wake_word_settings.timeout)
+
+        # Audio chunk buffer.
+        audio_bytes_to_buffer = int(
+            wake_word_settings.audio_seconds_to_buffer * 16000 * 2
+        )
+        audio_ring_buffer = b""
+
+        async def timestamped_stream() -> AsyncIterable[tuple[bytes, int]]:
+            """Yield audio with timestamps (milliseconds since start of stream)."""
+            nonlocal audio_ring_buffer
+
+            timestamp_ms = 0
+            async for chunk in stream:
+                yield chunk, timestamp_ms
+                timestamp_ms += (len(chunk) // 2) // 16  # milliseconds @ 16Khz
+
+                # Keeping audio right before wake word detection allows the
+                # voice command to be spoken immediately after the wake word.
+                if audio_bytes_to_buffer > 0:
+                    audio_ring_buffer += chunk
+                    if len(audio_ring_buffer) > audio_bytes_to_buffer:
+                        # A proper ring buffer would be far more efficient
+                        audio_ring_buffer = audio_ring_buffer[
+                            len(audio_ring_buffer) - audio_bytes_to_buffer :
+                        ]
+
+                if (wake_word_vad is not None) and (not wake_word_vad.process(chunk)):
+                    raise WakeWordTimeoutError(
+                        code="wake-word-timeout", message="Wake word was not detected"
+                    )
+
+        try:
+            # Detect wake word(s)
+            result = await self.wake_word_provider.async_process_audio_stream(
+                timestamped_stream()
+            )
+
+            if audio_ring_buffer:
+                # All audio kept from right before the wake word was detected as
+                # a single chunk.
+                audio_buffer.append(audio_ring_buffer)
+        except WakeWordTimeoutError:
+            _LOGGER.debug("Timeout during wake word detection")
+            raise
+        except Exception as src_error:
+            _LOGGER.exception("Unexpected error during wake-word-detection")
+            raise WakeWordDetectionError(
+                code="wake-stream-failed",
+                message="Unexpected error during wake-word-detection",
+            ) from src_error
+
+        _LOGGER.debug("wake-word-detection result %s", result)
+
+        if result is None:
+            wake_word_output: dict[str, Any] = {}
+        else:
+            if result.queued_audio:
+                # Add audio that was pending at detection
+                for chunk_ts in result.queued_audio:
+                    audio_buffer.append(chunk_ts[0])
```

Comment:
Wake word and STT share the same audio stream (async generator). In the Wyoming wake word code, audio chunks are pulled from this stream at the same time as detection events are coming from the wake word add-on (over TCP). 

It can happen that a detection event comes in when an audio chunk task is already pending. We can't put the chunk back into the stream for STT, and we don't want to drop the audio (since it could be the start of the voice command). So it gets slipped into the detection result and added to the buffer after the pre-detection wake word audio.